packages/pypy/build.sh

```diff
@@ -0,0 +1,267 @@
+_proot_checksums=d1eb20cb201e6df08d707023efb000623ff7c10d6574839d7bb42d0adba6b4da
+
+termux_step_pre_configure() {
+	if $TERMUX_ON_DEVICE_BUILD; then
+		termux_error_exit "Package '$TERMUX_PKG_NAME' is not safe for on-device builds."
+	fi
+
+	DOCKER_PULL="python $TERMUX_PKG_CACHEDIR/docker_pull.py"
+	UNDOCKER="python $TERMUX_PKG_CACHEDIR/undocker.py"
+	PROOT="$TERMUX_PKG_CACHEDIR/proot"
+
+	# Get docker_pull.py
+	termux_download \
+		$_docker_pull_url \
+		$TERMUX_PKG_CACHEDIR/docker_pull.py \
+		$_docker_pull_checksums
+
+	# Get undocker.py
+	termux_download \
+		$_undocker_url \
+		$TERMUX_PKG_CACHEDIR/undocker.py \
+		$_undocker_checksums
+	
+	# Get proot
+	termux_download \
+		$_proot_url \
+		$TERMUX_PKG_CACHEDIR/proot \
+		$_proot_checksums
+	
+	chmod +x $TERMUX_PKG_CACHEDIR/proot
+	
+	# Pick up host platform arch.
+	HOST_ARCH=$TERMUX_ARCH
+	if [ $TERMUX_ARCH = "arm" ]; then
+		HOST_ARCH="i686"
+	elif [ $TERMUX_ARCH = "aarch64" ]; then
+		HOST_ARCH="x86_64"
+	fi
+
+	# Get host platform rootfs tar if needed.
+	if [ ! -f "$TERMUX_PKG_CACHEDIR/kcubeterm_termux_$HOST_ARCH.tar" ]; then
+		(
+			cd $TERMUX_PKG_CACHEDIR
+			$DOCKER_PULL kcubeterm/termux:$HOST_ARCH
+			mv kcubeterm_termux.tar kcubeterm_termux_$HOST_ARCH.tar
+		)
+	fi
+
+	# Get target platform rootfs tar if needed.
+	if [ $HOST_ARCH != $TERMUX_ARCH ]; then
+		if [ ! -f "$TERMUX_PKG_CACHEDIR/kcubeterm_termux_$TERMUX_ARCH.tar" ]; then
+			(
+				cd $TERMUX_PKG_CACHEDIR
+				$DOCKER_PULL kcubeterm/termux:$TERMUX_ARCH
+				mv kcubeterm_termux.tar kcubeterm_termux_$TERMUX_ARCH.tar
+			)
+		fi
+	fi
+}
+
+termux_step_configure() {
+	PYPY_USESSION_DIR=$TERMUX_ANDROID_HOME/tmp
+	PYPY_SRC_DIR=$TERMUX_ANDROID_HOME/src
+
+	# Bootstrap a proot rootfs for the host platform
+	HOST_ROOTFS_BASE=$TERMUX_PKG_TMPDIR/host-rootfs
+	cat "$TERMUX_PKG_CACHEDIR/kcubeterm_termux_$HOST_ARCH.tar" | $UNDOCKER -o $HOST_ROOTFS_BASE
+
+	# Add build dependicies for pypy on the host platform rootfs
+	# Build essential
+	BUILD_DEP="binutils binutils-gold clang file patch pkg-config python2"
+	# Build dependencies for pypy
+	BUILD_DEP+=" gdbm libandroid-posix-semaphore libandroid-support libbz2"
+	BUILD_DEP+=" libcrypt libexpat libffi liblzma libsqlite libxml2 ncurses"
+	BUILD_DEP+=" ncurses-ui-libs openssl readline zlib tk xorgproto dash"
+	BUILD_DEO+=" ndk-sysroot"
+
+	# Environment variables for termux
+	TERMUX_RUNTIME_ENV_VARS="ANDROID_DATA=/data
+			ANDROID_ROOT=/system
+			HOME=$TERMUX_ANDROID_HOME
+			LANG=en_US.UTF-8
+			PATH=$TERMUX_PREFIX/bin
+			PREFIX=$TERMUX_PREFIX
+			TMPDIR=$TERMUX_PREFIX/tmp
+			TZ=UTC"
+
+	PROOT_HOST="env -i PROOT_NO_SECCOMP=1
+						$TERMUX_RUNTIME_ENV_VARS
+						$PROOT 
+						-b /proc -b /dev -b /sys
+						-b /:$TERMUX_ANDROID_HOME/ndk-rootfs
+						-b $TERMUX_PKG_SRCDIR:$PYPY_SRC_DIR
+						-w $TERMUX_ANDROID_HOME
+						-r $HOST_ROOTFS_BASE/"
+	$PROOT_HOST update-static-dns
+	$PROOT_HOST pkg install -yq $BUILD_DEP
```

Comment:
Since we are installing the dependencies inside the proot environment, should we skip TERMUX_BUILD_DEPENDS to avoid duplicate work? Unless there's a potential on-device build support, I dont see the need to include both.

Reply:
Due to the linker diffrence between glibc and bionic, PyPy's cross-build should perform the translation under a x86(_64) proot environment rather than under the system environment. And during the translation, we have to execute the built binary under another proot environment such as an aarch64 one or a x86(_64).one. But actually the x86(_64) proot environment is only for translation, finally we `make` using the Makefile under the system environment, so the dependencies may be necessary.
Maybe we could perform all the operations in the x86(_64) proot enviroment later if `clang-13` in termux could actually cross-compile for arm. Now, `pkg install build-essential ndk-sysroot ndk-multilib` and then `clang --target=aarch64-unknown-linux-android` says `libunwind.a` not found on a x86_64 device...

Reply:
Ok I understand pypy's unique build system that we have to tackle this differently.
What I meant is can we mount `/data` or specifically `/data/data/com.termux/files/usr` in proot environment? I believe we can skip the `pkg install` step if prefix is mounted.

Reply:
Emmm... When build an aarch64 binary of pypy, the proot environment performing the translation is a x86_64 environment, but `$TERMUX_PREFIX` contains the files whose arch is aarch64. The step `pkg install` seemds hard to skip...

Reply:
Now `clang` can cross compile under a x86(_64) proot environment. I will try to refactor this script.